src/core/ext/filters/client_channel/lb_policy/xds/xds_load_balancer_api.cc

```diff
@@ -18,62 +18,167 @@
-  if (GPR_UNLIKELY(!pb_decode(stream, grpc_lb_v1_Server_fields, server))) {
-    gpr_free(server);
-    gpr_log(GPR_ERROR, "nanopb error: %s", PB_GET_ERROR(stream));
-    return false;
+  // Populate grpc_resolved_address.
+  grpc_resolved_address addr;
+  char* address_str = static_cast<char*>(gpr_malloc(address_strview.size + 1));
+  memcpy(address_str, address_strview.data, address_strview.size);
+  address_str[address_strview.size] = '\0';
+  grpc_string_to_sockaddr(&addr, address_str, port);
+  gpr_free(address_str);
+  // Append the address to the list.
+  list->emplace_back(addr, nullptr);
+  return GRPC_ERROR_NONE;
+}
+
+grpc_error* LocalityParse(const XdsLocalityLbEndpoints* locality_lb_endpoints,
+                          XdsLocalityInfo* locality_info) {
+  upb::Arena arena;
+  // Parse locality name.
+  const XdsLocality* locality =
+      envoy_api_v2_endpoint_LocalityLbEndpoints_locality(locality_lb_endpoints);
+  locality_info->locality_name = MakeRefCounted<XdsLocalityName>(
+      UniquePtr<char>(
+          gpr_strdup(envoy_api_v2_core_Locality_region(locality).data)),
+      UniquePtr<char>(
+          gpr_strdup(envoy_api_v2_core_Locality_zone(locality).data)),
+      UniquePtr<char>(
+          gpr_strdup(envoy_api_v2_core_Locality_sub_zone(locality).data)));
+  // Parse the addresses.
+  size_t size;
+  const XdsLbEndpoint* const* lb_endpoints =
+      envoy_api_v2_endpoint_LocalityLbEndpoints_lb_endpoints(
+          locality_lb_endpoints, &size);
+  for (size_t i = 0; i < size; ++i) {
+    grpc_error* error = ServerAddressParseAndAppend(lb_endpoints[i],
+                                                    &locality_info->serverlist);
+    if (error != GRPC_ERROR_NONE) return error;
   }
-  dec_arg->serverlist->servers[dec_arg->decoding_idx++] = server;
-  return true;
-}
-
-xds_grpclb_request* xds_grpclb_request_create(const char* lb_service_name) {
-  xds_grpclb_request* req =
-      static_cast<xds_grpclb_request*>(gpr_malloc(sizeof(xds_grpclb_request)));
-  req->has_client_stats = false;
-  req->has_initial_request = true;
-  req->initial_request.has_name = true;
-  // GCC warns (-Wstringop-truncation) because the destination
-  // buffer size is identical to max-size, leading to a potential
-  // char[] with no null terminator.  nanopb can handle it fine,
-  // and parantheses around strncpy silence that compiler warning.
-  (strncpy(req->initial_request.name, lb_service_name,
-           XDS_SERVICE_NAME_MAX_LENGTH));
-  return req;
+  // Parse the lb_weight and priority.
+  const google_protobuf_UInt32Value* lb_weight =
+      envoy_api_v2_endpoint_LocalityLbEndpoints_load_balancing_weight(
+          locality_lb_endpoints);
+  // If LB weight is not specified, the default weight 0 is used, which means
+  // this locality is assigned no load.
+  locality_info->lb_weight =
+      lb_weight != nullptr ? google_protobuf_UInt32Value_value(lb_weight) : 0;
```

Comment:
Hmm, I realized I am not providing has_foo() accessors for proto3 submessage fields. That seems like it would be better than checking them against `nullptr`. I am trying to figure out if I should return a default instance here instead. Do you have any opinion about that?